test/conformance/errorcondition_test.go

```diff
@@ -132,65 +126,90 @@ func TestContainerExitingMsg(t *testing.T) {
 		errorLog = "Crashed..."
 	)
 
-	t.Logf("Creating a new Configuration %s", names.Image)
-
-	// This probe is crucial for having a race free conformance test. It will prevent the
-	// pod from becoming ready intermittently.
-	probe := &corev1.Probe{
-		Handler: corev1.Handler{
-			HTTPGet: &corev1.HTTPGetAction{},
+	tests := []struct {
+		Name           string
+		ReadinessProbe *corev1.Probe
+	}{{
+		Name: "http",
+		ReadinessProbe: &corev1.Probe{
+			Handler: corev1.Handler{
+				HTTPGet: &corev1.HTTPGetAction{},
+			},
 		},
-	}
-	if _, err := test.CreateConfiguration(t, clients, names, &test.Options{ReadinessProbe: probe}); err != nil {
-		t.Fatalf("Failed to create configuration %s: %v", names.Config, err)
-	}
-	defer test.TearDown(clients, names)
-	test.CleanupOnInterrupt(func() { test.TearDown(clients, names) })
+	}, {
+		Name: "tcp",
+		ReadinessProbe: &corev1.Probe{
+			Handler: corev1.Handler{
+				TCPSocket: &corev1.TCPSocketAction{},
+			},
+		},
+	}}
 
-	t.Log("When the containers keep crashing, the Configuration should have error status.")
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.Name, func(t *testing.T) {
+			t.Parallel()
+			clients := setup(t)
 
-	err := test.WaitForConfigurationState(clients.ServingClient, names.Config, func(r *v1alpha1.Configuration) (bool, error) {
-		cond := r.Status.GetCondition(v1alpha1.ConfigurationConditionReady)
-		if cond != nil && !cond.IsUnknown() {
-			if strings.Contains(cond.Message, errorLog) && cond.IsFalse() {
-				return true, nil
+			names := test.ResourceNames{
+				Config: test.ObjectNameForTest(t),
+				Image:  "failing",
 			}
-			t.Logf("Reason: %s ; Message: %s ; Status: %s", cond.Reason, cond.Message, cond.Status)
-			return true, fmt.Errorf("The configuration %s was not marked with expected error condition (Reason=\"%s\", Message=\"%s\", Status=\"%s\"), but with (Reason=\"%s\", Message=\"%s\", Status=\"%s\")",
-				names.Config, containerMissing, errorLog, "False", cond.Reason, cond.Message, cond.Status)
-		}
-		return false, nil
-	}, "ConfigContainersCrashing")
 
-	if err != nil {
-		t.Fatalf("Failed to validate configuration state: %s", err)
-	}
+			t.Logf("Creating a new Configuration %s", names.Image)
 
-	revisionName, err := getRevisionFromConfiguration(clients, names.Config)
-	if err != nil {
-		t.Fatalf("Failed to get revision from configuration %s: %v", names.Config, err)
-	}
+			if _, err := test.CreateConfiguration(t, clients, names, &test.Options{ReadinessProbe: tt.ReadinessProbe}); err != nil {
+				t.Fatalf("Failed to create configuration %s: %v", names.Config, err)
+			}
+			defer test.TearDown(clients, names)
+			test.CleanupOnInterrupt(func() { test.TearDown(clients, names) })
+
+			t.Log("When the containers keep crashing, the Configuration should have error status.")
+
```

Comment:
I'd include the test name (http vs tcp) in the logging as they are running n parallel.

Reply:
The output does not get printed until the test is completed. Thus, the log messages are not mixed up. From the current output, I think it's clear which test is run, but feel free to disagree and suggest other improvements. 

Here's what a sample output: 
```
=== RUN   TestContainerExitingMsg
=== RUN   TestContainerExitingMsg/http
=== PAUSE TestContainerExitingMsg/http
=== RUN   TestContainerExitingMsg/tcp
=== PAUSE TestContainerExitingMsg/tcp
=== CONT  TestContainerExitingMsg/http
=== CONT  TestContainerExitingMsg/tcp
--- PASS: TestContainerExitingMsg (0.00s)
    --- PASS: TestContainerExitingMsg/tcp (121.49s)
        errorcondition_test.go:157: Creating a new Configuration failing
        util.go:34: resource (test.ResourceObjects) {
             Route: (*v1alpha1.Route)(<nil>),
             Config: (*v1alpha1.Configuration)(0xc0001d4400)({
              TypeMeta: (v1.TypeMeta) &TypeMeta{Kind:,APIVersion:,},
              ObjectMeta: (v1.ObjectMeta) &ObjectMeta{Name:container-exiting-msg-tcp-rlxhqmkq,GenerateName:,Namespace:serving-tests,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,},
              Spec: (v1alpha1.ConfigurationSpec) {
               DeprecatedGeneration: (int64) 0,
               Build: (*v1alpha1.RawExtension)(<nil>),
               RevisionTemplate: (v1alpha1.RevisionTemplateSpec) &ObjectMeta{Name:,GenerateName:,Namespace:,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,}
              },
              Status: (v1alpha1.ConfigurationStatus) {
               Status: (v1alpha1.Status) {
                ObservedGeneration: (int64) 0,
                Conditions: (v1alpha1.Conditions) <nil>
               },
               ConfigurationStatusFields: (v1alpha1.ConfigurationStatusFields) {
                LatestReadyRevisionName: (string) "",
                LatestCreatedRevisionName: (string) ""
               }
              }
             }),
             Service: (*v1alpha1.Service)(<nil>),
             Revision: (*v1alpha1.Revision)(<nil>)
            }
        errorcondition_test.go:165: When the containers keep crashing, the Configuration should have error status.
        errorcondition_test.go:189: When the containers keep crashing, the revision should have error status.
        errorcondition_test.go:206: When the revision has error condition, logUrl should be populated.
    --- PASS: TestContainerExitingMsg/http (122.53s)
        errorcondition_test.go:157: Creating a new Configuration failing
        util.go:34: resource (test.ResourceObjects) {
             Route: (*v1alpha1.Route)(<nil>),
             Config: (*v1alpha1.Configuration)(0xc00052e400)({
              TypeMeta: (v1.TypeMeta) &TypeMeta{Kind:,APIVersion:,},
              ObjectMeta: (v1.ObjectMeta) &ObjectMeta{Name:container-exiting-msg-http-gveihupq,GenerateName:,Namespace:serving-tests,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,},
              Spec: (v1alpha1.ConfigurationSpec) {
               DeprecatedGeneration: (int64) 0,
               Build: (*v1alpha1.RawExtension)(<nil>),
               RevisionTemplate: (v1alpha1.RevisionTemplateSpec) &ObjectMeta{Name:,GenerateName:,Namespace:,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,}
              },
              Status: (v1alpha1.ConfigurationStatus) {
               Status: (v1alpha1.Status) {
                ObservedGeneration: (int64) 0,
                Conditions: (v1alpha1.Conditions) <nil>
               },
               ConfigurationStatusFields: (v1alpha1.ConfigurationStatusFields) {
                LatestReadyRevisionName: (string) "",
                LatestCreatedRevisionName: (string) ""
               }
              }
             }),
             Service: (*v1alpha1.Service)(<nil>),
             Revision: (*v1alpha1.Revision)(<nil>)
            }
        errorcondition_test.go:165: When the containers keep crashing, the Configuration should have error status.
        errorcondition_test.go:189: When the containers keep crashing, the revision should have error status.
        errorcondition_test.go:206: When the revision has error condition, logUrl should be populated.
PASS
ok  	github.com/knative/serving/test/conformance	122.948s

```

Reply:
Yes.  They all look good when they pass.   However, when something goes wrong, we may need to match the logs to figure out which message has failed/gone missing.   It probably is worth logging the extra info.   WDYT?

Reply:
Adding a sample output for a failed test. For http, I made it always fail on validation configuration state [Line 180](https://github.com/knative/serving/blob/77abfd961a6a1fb7116e915ac24bf7fea0e7363e/test/conformance/errorcondition_test.go#L180) by checking the test name. Similarly, made tcp to fail on validation on the revision state [Line 202](https://github.com/knative/serving/blob/77abfd961a6a1fb7116e915ac24bf7fea0e7363e/test/conformance/errorcondition_test.go#L202). 

In my opinion, I think the current logging is good enough. Is there a particular failure case where you think it would mix up the log/error messages? 

<details>
  <summary>Sample failed output</summary>
  
  ```
=== RUN   TestContainerExitingMsg
=== RUN   TestContainerExitingMsg/http
=== PAUSE TestContainerExitingMsg/http
=== RUN   TestContainerExitingMsg/tcp
=== PAUSE TestContainerExitingMsg/tcp
=== CONT  TestContainerExitingMsg/http
=== CONT  TestContainerExitingMsg/tcp
--- FAIL: TestContainerExitingMsg (0.00s)
    --- FAIL: TestContainerExitingMsg/http (18.22s)
        errorcondition_test.go:157: Creating a new Configuration failing
        util.go:34: resource (test.ResourceObjects) {
             Route: (*v1alpha1.Route)(<nil>),
             Config: (*v1alpha1.Configuration)(0xc00058a400)({
              TypeMeta: (v1.TypeMeta) &TypeMeta{Kind:,APIVersion:,},
              ObjectMeta: (v1.ObjectMeta) &ObjectMeta{Name:container-exiting-msg-http-ztkwidvb,GenerateName:,Namespace:serving-tests,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,},
              Spec: (v1alpha1.ConfigurationSpec) {
               DeprecatedGeneration: (int64) 0,
               Build: (*v1alpha1.RawExtension)(<nil>),
               RevisionTemplate: (v1alpha1.RevisionTemplateSpec) &ObjectMeta{Name:,GenerateName:,Namespace:,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,}
              },
              Status: (v1alpha1.ConfigurationStatus) {
               Status: (v1alpha1.Status) {
                ObservedGeneration: (int64) 0,
                Conditions: (v1alpha1.Conditions) <nil>
               },
               ConfigurationStatusFields: (v1alpha1.ConfigurationStatusFields) {
                LatestReadyRevisionName: (string) "",
                LatestCreatedRevisionName: (string) ""
               }
              }
             }),
             Service: (*v1alpha1.Service)(<nil>),
             Revision: (*v1alpha1.Revision)(<nil>)
            }
        errorcondition_test.go:166: When the containers keep crashing, the Configuration should have error status.
        errorcondition_test.go:182: Failed to validate configuration state: %!s(<nil>)
    --- FAIL: TestContainerExitingMsg/tcp (18.28s)
        errorcondition_test.go:157: Creating a new Configuration failing
        util.go:34: resource (test.ResourceObjects) {
             Route: (*v1alpha1.Route)(<nil>),
             Config: (*v1alpha1.Configuration)(0xc0005ce400)({
              TypeMeta: (v1.TypeMeta) &TypeMeta{Kind:,APIVersion:,},
              ObjectMeta: (v1.ObjectMeta) &ObjectMeta{Name:container-exiting-msg-tcp-eznuajeg,GenerateName:,Namespace:serving-tests,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,},
              Spec: (v1alpha1.ConfigurationSpec) {
               DeprecatedGeneration: (int64) 0,
               Build: (*v1alpha1.RawExtension)(<nil>),
               RevisionTemplate: (v1alpha1.RevisionTemplateSpec) &ObjectMeta{Name:,GenerateName:,Namespace:,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,}
              },
              Status: (v1alpha1.ConfigurationStatus) {
               Status: (v1alpha1.Status) {
                ObservedGeneration: (int64) 0,
                Conditions: (v1alpha1.Conditions) <nil>
               },
               ConfigurationStatusFields: (v1alpha1.ConfigurationStatusFields) {
                LatestReadyRevisionName: (string) "",
                LatestCreatedRevisionName: (string) ""
               }
              }
             }),
             Service: (*v1alpha1.Service)(<nil>),
             Revision: (*v1alpha1.Revision)(<nil>)
            }
        errorcondition_test.go:166: When the containers keep crashing, the Configuration should have error status.
        errorcondition_test.go:190: When the containers keep crashing, the revision should have error status.
        errorcondition_test.go:204: Failed to validate revision state: %!s(<nil>)
FAIL
```
</details>

Reply:
I doubt that tcp/http "sub-tests" will ever run in parallel within the same test unit.   So, I'd keep the `t.Parallel()` at line 120 (at the beginning of the function) and delete the `t.Parallel()` at line 151 (inside the for-loop).
As far as the log output, please try increasing the count and check the log output.  e.g. ```go test -v -tags=e2e -count=8 -cpu=8 ./test/conformance -run ^TestContainerExitingMsg$ --timeout 30m```.    I agree that it is not easy to cause the log to mix up as output is printed when a test is complete.

Reply:
I think the subtests are run in parallel within the test unit. 

For the current test i.e. `t.Parallel()` within each subtests, the output is 
```
=== RUN   TestContainerExitingMsg
=== RUN   TestContainerExitingMsg/http
=== PAUSE TestContainerExitingMsg/http
=== RUN   TestContainerExitingMsg/tcp
=== PAUSE TestContainerExitingMsg/tcp
=== CONT  TestContainerExitingMsg/http
=== CONT  TestContainerExitingMsg/tcp
<A single wait before it prints the test result>
...
```

After moving `t.Parallel()` to the beginning

```
=== RUN   TestContainerExitingMsg
=== PAUSE TestContainerExitingMsg
=== CONT  TestContainerExitingMsg
=== RUN   TestContainerExitingMsg/http 
<There's a wait before it prints the next line> 
=== RUN   TestContainerExitingMsg/tcp
<Another wait before it prints the test result>
...
```

<details>
  <summary>Full output when `t.Parallel()` is moved to the beginning of the test</summary>

```
=== RUN   TestContainerExitingMsg
=== PAUSE TestContainerExitingMsg
=== CONT  TestContainerExitingMsg
=== RUN   TestContainerExitingMsg/http
=== RUN   TestContainerExitingMsg/tcp
--- PASS: TestContainerExitingMsg (164.48s)
    --- PASS: TestContainerExitingMsg/http (122.33s)
        errorcondition_test.go:158: Creating a new Configuration failing
        util.go:34: resource (test.ResourceObjects) {
             Route: (*v1alpha1.Route)(<nil>),
             Config: (*v1alpha1.Configuration)(0xc000073400)({
              TypeMeta: (v1.TypeMeta) &TypeMeta{Kind:,APIVersion:,},
              ObjectMeta: (v1.ObjectMeta) &ObjectMeta{Name:container-exiting-msg-http-smorkmba,GenerateName:,Namespace:serving-tests,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,},
              Spec: (v1alpha1.ConfigurationSpec) {
               DeprecatedGeneration: (int64) 0,
               Build: (*v1alpha1.RawExtension)(<nil>),
               RevisionTemplate: (v1alpha1.RevisionTemplateSpec) &ObjectMeta{Name:,GenerateName:,Namespace:,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,}
              },
              Status: (v1alpha1.ConfigurationStatus) {
               Status: (v1alpha1.Status) {
                ObservedGeneration: (int64) 0,
                Conditions: (v1alpha1.Conditions) <nil>
               },
               ConfigurationStatusFields: (v1alpha1.ConfigurationStatusFields) {
                LatestReadyRevisionName: (string) "",
                LatestCreatedRevisionName: (string) ""
               }
              }
             }),
             Service: (*v1alpha1.Service)(<nil>),
             Revision: (*v1alpha1.Revision)(<nil>)
            }
        errorcondition_test.go:166: When the containers keep crashing, the Configuration should have error status.
        errorcondition_test.go:190: When the containers keep crashing, the revision should have error status.
        errorcondition_test.go:207: When the revision has error condition, logUrl should be populated.
    --- PASS: TestContainerExitingMsg/tcp (42.16s)
        errorcondition_test.go:158: Creating a new Configuration failing
        util.go:34: resource (test.ResourceObjects) {
             Route: (*v1alpha1.Route)(<nil>),
             Config: (*v1alpha1.Configuration)(0xc00054e400)({
              TypeMeta: (v1.TypeMeta) &TypeMeta{Kind:,APIVersion:,},
              ObjectMeta: (v1.ObjectMeta) &ObjectMeta{Name:container-exiting-msg-tcp-cybbzcpn,GenerateName:,Namespace:serving-tests,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,},
              Spec: (v1alpha1.ConfigurationSpec) {
               DeprecatedGeneration: (int64) 0,
               Build: (*v1alpha1.RawExtension)(<nil>),
               RevisionTemplate: (v1alpha1.RevisionTemplateSpec) &ObjectMeta{Name:,GenerateName:,Namespace:,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,}
              },
              Status: (v1alpha1.ConfigurationStatus) {
               Status: (v1alpha1.Status) {
                ObservedGeneration: (int64) 0,
                Conditions: (v1alpha1.Conditions) <nil>
               },
               ConfigurationStatusFields: (v1alpha1.ConfigurationStatusFields) {
                LatestReadyRevisionName: (string) "",
                LatestCreatedRevisionName: (string) ""
               }
              }
             }),
             Service: (*v1alpha1.Service)(<nil>),
             Revision: (*v1alpha1.Revision)(<nil>)
            }
        errorcondition_test.go:166: When the containers keep crashing, the Configuration should have error status.
        errorcondition_test.go:190: When the containers keep crashing, the revision should have error status.
        errorcondition_test.go:207: When the revision has error condition, logUrl should be populated.
PASS
```
</details>

Thanks for suggesting a way to see if the log output can be mixed up. I still running `go test -v -tags=e2e -count=8 -cpu=8 ./test/conformance -run ^TestContainerExitingMsg$ --timeout 30m` in the background to see if I can at least see a case where the log output is mixed up, but I haven't seen the case where the log is mixed up. Since it is not easy to get to this state, I think the current logging is enough to determine which of the http/tcp case fails.

Reply:
Yes. Subtests do indeed run in parallel.